sdks/go/test/integration/io/xlang/bigtable/bigtable_test.go

```diff
@@ -0,0 +1,177 @@
+	beam.RegisterDoFn(&RowToMutationFn{})
+
+	register.DoFn2x0[[]byte, func(xlangbt.Row)](&CreateTestRowsFn{})
+	register.DoFn2x0[xlangbt.Row, func(mutation bigtableio.Mutation)](&RowToMutationFn{})
+	register.Emitter1[bigtableio.Mutation]()
+	register.Emitter1[xlangbt.Row]()
+}
+
+type CreateTestRowsFn struct {
+}
+
+type RowToMutationFn struct {
+}
+
+func (fn *RowToMutationFn) ProcessElement(r xlangbt.Row, emit func(mutation bigtableio.Mutation)) {
+	mut := bigtableio.NewMutation(string(r.Key))
+	for cf, cols := range r.Column_families {
+		for col, vals := range cols {
+			for _, val := range vals {
+				mut.Set(cf, col, bt.Timestamp(val.Timestamp_micros), val.Value)
+			}
+		}
+	}
+	emit(*mut)
+}
+
+func createTestRows() []xlangbt.Row {
+	rows := make([]xlangbt.Row, 100)
+
+	for i := 0; i < 100; i++ {
+		key := strconv.FormatInt(int64(i), 10)
+		row := xlangbt.Row{
+			Key: []byte(key),
+		}
+		row.AddCell("cf1", "q1", []byte(strconv.FormatInt(int64(i)*1000, 10)), int64(i)*1000)
+
+		rows[i] = row
+	}
+	return rows
+}
+
+func (fn *CreateTestRowsFn) ProcessElement(_ []byte, emit func(row xlangbt.Row)) {
+	for _, r := range createTestRows() {
+		emit(r)
+	}
+}
+
+func checkFlags(t *testing.T) {
+	if *integration.BigtableInstance == "" {
+		t.Skip("No Bigtable instance provided.")
+	}
+}
+
+func WritePipeline(project, instance, table string) *beam.Pipeline {
+	p := beam.NewPipeline()
+	s := p.Root()
+
+	rows := beam.ParDo(s, &CreateTestRowsFn{}, beam.Impulse(s))
+	muts := beam.ParDo(s, &RowToMutationFn{}, rows)
+	bigtableio.Write(s, project, instance, table, muts)
+	return p
+}
+
+func ReadPipeline(project, instance, table, expansionAddr string) *beam.Pipeline {
+	p := beam.NewPipeline()
+	s := p.Root()
+
+	rowsFromBt := xlangbt.Read(s, project, instance, table, xlangbt.ReadExpansionAddr(expansionAddr))
+
+	rows := beam.ParDo(s, &CreateTestRowsFn{}, beam.Impulse(s))
+
+	passert.Equals(s, rowsFromBt, rows)
+	return p
+}
+
+var instanceRegex = regexp.MustCompile("projects/(?P<project>[^/]+)/instances/(?P<instance>.+)")
+
+func createTempTable(admin *bt.AdminClient) string {
+	tableName := fmt.Sprintf("go_btio_it_temp_%v", time.Now().UnixNano())
+	err := admin.CreateTableFromConf(context.Background(), &bt.TableConf{
+		TableID: tableName,
+		Families: map[string]bt.GCPolicy{
+			"cf1": bt.NoGcPolicy(),
+		},
+	})
+	if err != nil {
+		panic(err)
+	}
+
+	return tableName
+}
+
+func deleteTempTable(admin *bt.AdminClient, table string) {
+	_ = admin.DeleteTable(context.Background(), table)
+}
+
+func TestBigtableIO_BasicWriteRead(t *testing.T) {
+	integration.CheckFilters(t)
+	checkFlags(t)
+
+	instancePath := *integration.BigtableInstance
+	matches := instanceRegex.FindStringSubmatch(instancePath)
+	project := matches[1]
+	instance := matches[2]
+
+	admin, err := bt.NewAdminClient(context.Background(), project, instance)
+	if err != nil {
+		panic(err)
+	}
+
+	table := createTempTable(admin)
+	defer deleteTempTable(admin, table)
```

Comment:
Prefer t.Cleanup over defer in test cases 

https://pkg.go.dev/testing#T.Cleanup

Reply:
👍